### DISCUSSION TITLE
Example of git-refs

### DISCUSSION THREAD

**Discussion by samuel-p** [renovatebot/renovate#8402] [Request Help] [answered]
Is there an example of how to use `git-refs` datasource? 

I want to get updates for an git repository, to merge it into my current project. I also don't know if it is even the right datasource to achieve this.

**Comment by rarkins** [comment]
Are you using the regex manager? I'm not sure we have a use case for git-refs in the codebase currently.

**Comment by samuel-p** [comment]
Yes I'm using the regex manager or at least tried to. I want to build a Docker image for a git repository and rebuild it if the repository gets updated. Here's the config I have for now:

```
{
  "regexManagers": [{
      "fileMatch": ["^Dockerfile$"],
      "matchStrings": ["\\s+ENV\\s*COMMIT_HASH=\"(?<currentVersion>.*?)\"\\s+"],
      "depNameTemplate": "<git-url>",
      "versioningTemplate": "git",
      "datasourceTemplate": "git-refs"
  }]
}
```

**Reply by rarkins** [reply]
The commit hash would need to be the `currentDigest`, not `currentVersion`.

**Comment by samuel-p** [comment]
If I change `currentVersion` to `currentDigest` renovatebot creates the following issue:

> There is an error with this repository's Renovate configuration that needs to be fixed. As a precaution, Renovate will stop PRs until it is resolved.
> File: `renovate.json`
Error type: The renovate configuration file contains some invalid settings
Message: `Regex Managers must contain currentValueTemplate configuration or regex group named currentValue`

**Reply by rarkins** [reply]
Can you try setting `currentValueTemplate` to be the name of the default branch? e.g. `master` or `main`? If that fails, try setting it to `null`. It may just be necessary for validation and not the lookup itself.

**Comment by samuel-p** [comment]
For both, `main` and `null` i get the following issue:

> There is an error with this repository's Renovate configuration that needs to be fixed. As a precaution, Renovate will stop PRs until it is resolved.
> File: `renovate.json`
Error type: The renovate configuration file contains some invalid settings
Message: `Invalid configuration option: regexManagers[0].currentValueTemplate, Regex Manager contains disallowed fields: currentValueTemplate`

**Comment by samuel-p** [comment, marked as answer]
If I add `currentValue` as an empty group to the regex and change it to `\s+ENV\s*MORTY_VERSION="(?<currentValue>)(?<currentDigest>.*?)"\s+` it works.

The full config now looks like this:

```
{
  "regexManagers": [{
      "fileMatch": ["^Dockerfile$"],
      "matchStrings": ["\s+ENV\s*COMMIT_HASH="(?<currentValue>)(?<currentDigest>.*?)"\s+"],
      "depNameTemplate": "<git-url>",
      "versioningTemplate": "git",
      "datasourceTemplate": "git-refs"
  }]
}
```

**Comment by MPV** [comment]
Hope you don't mind me borrowing the discussion for an almost identical use case 🙏 

I have this regex manager code (thanks for the currentValue/currentDigest hack above, it made me get `updates` values):
- https://github.com/MPV/dotfiles/blob/d5e01c47a8e86247a1da2a577f5f12aac91cf7ac/renovate.json#L7-L18
```json
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": ["^\\.plugin-versions$"],
      "matchStrings": [
        "(?<depName>\\S+)\\W*(?<packageName>https:\\/\\/github.com\\/[\\S]+)\\W*(?<currentValue>)(?<currentDigest>[a-f0-9]{7,40})"
      ],
      "autoReplaceStringTemplate": "{{{depName}}} {{{packageName}}} {{{newDigest}}}",
      "datasourceTemplate": "git-refs",
      "versioningTemplate": "git"
    }
  ],
```

And I want Renovate to upgrade the commit sha/digests in the end of this file:
- https://github.com/MPV/dotfiles/blob/master/.plugin-versions

With the regex proven like this:
- https://regex101.com/r/YaruP5/2

And I get job logs finding updates:
- https://developer.mend.io/github/MPV/dotfiles/-/job/9b8f5f8c-9915-4a0f-acc2-607f1b5ccbc8
```json
          {
            "depName": "pluto",
            "packageName": "https://github.com/FairwindsOps/asdf-pluto.git",
            "currentDigest": "e30adc6",
            "datasource": "git-refs",
            "versioning": "git",
            "replaceString": "pluto                https://github.com/FairwindsOps/asdf-pluto.git             e30adc6",
            "updates": [
              {
                "updateType": "digest",
                "newDigest": "1e24ad5801a02eecceb46d32d2622b9f54de4535",
                "branchName": "renovate/pluto-digest"
              }
            ],
            "warnings": []
          },
```
And there are upgrades available:
- https://github.com/FairwindsOps/asdf-pluto/compare/e30adc6...master

But I get no PRs. Any ideas?

**Reply by MPV** [reply]
If it helps, let's try comparing to my `git-submodules` PRs that work fine, i.e:
- https://github.com/MPV/dotfiles/pull/75

There I get this in the renovate logs instead:
```json
    {
      "branchName": "renovate/asdf-plugins-pluto-digest",
      "prNo": 75,
      "prTitle": "Update .asdf/plugins/pluto digest to 1e24ad5",
      "upgrades": [
        {
          "datasource": "git-refs",
          "depName": ".asdf/plugins/pluto",
          "displayPending": "",
          "currentValue": "master",
          "currentDigest": "e30adc62801b7656342800599a72be3ee4af65a0",
          "newValue": "master",
          "newDigest": "1e24ad5801a02eecceb46d32d2622b9f54de4535",
          "packageFile": ".gitmodules",
          "updateType": "digest",
          "packageName": "https://github.com/FairwindsOps/asdf-pluto.git"
        }
      ]
    },
```

So comparing them, we see these differences:
```diff
 {
-  "branchName": "renovate/asdf-plugins-pluto-digest",
-  "prNo": 75,
-  "prTitle": "Update .asdf/plugins/pluto digest to 1e24ad5",
+  "branchName": "renovate/pluto-digest",
+  "prNo": null,
+  "prTitle": "Update pluto digest to 1e24ad5",
   "upgrades": [
     {
       "datasource": "git-refs",
-      "depName": ".asdf/plugins/pluto",
+      "depName": "pluto",
       "displayPending": "",
-      "currentValue": "master",
-      "currentDigest": "e30adc62801b7656342800599a72be3ee4af65a0",
-      "newValue": "master",
+      "currentDigest": "e30adc6",
       "newDigest": "1e24ad5801a02eecceb46d32d2622b9f54de4535",
-      "packageFile": ".gitmodules",
+      "packageFile": ".plugin-versions",
       "updateType": "digest",
       "packageName": "https://github.com/FairwindsOps/asdf-pluto.git"
     }
```

While I maybe could go the way rarkins suggested and set:
> Can you try setting `currentValueTemplate` to be the name of the default branch? e.g. `master` or `main`? If that fails, try setting it to `null`. It may just be necessary for validation and not the lookup itself.

...that might not always work, as some of the git repos I depend on have different default branches (i.e. main vs master).

Could this be possible to solve anyway, and how?

**Reply by MPV** [reply]
An now suddenly it started working: 🎉
- https://github.com/MPV/dotfiles/pull/100
